### ISSUE TITLE
Requesting help for ThinQ v1 air purifier support

### ISSUE THREAD

**Issue by zvuc** [nVuln/homebridge-lg-thinq#84]
**Which device do you want to add? Please describe:**

LG Air purifier AS128VWA (2018?)

**Describe the solution you'd like:**

I want basic control support for Fan speed, turning on/off, observe air quality sensor data.

**Device Data:**

> (node:4292) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'type' of undefined
    at DeviceModel.decodeMonitor (/homebridge/node_modules/homebridge-lg-thinq/src/lib/DeviceModel.ts:204:30)
    at Function.transform (/homebridge/node_modules/homebridge-lg-thinq/src/v1/helper.ts:33:47)
    at ThinQ.pollMonitor (/homebridge/node_modules/homebridge-lg-thinq/src/lib/ThinQ.ts:163:21)
    at LGThinQHomebridgePlatform.discoverDevices (/homebridge/node_modules/homebridge-lg-thinq/src/platform.ts:100:51)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at HomebridgeAPI.<anonymous> (/homebridge/node_modules/homebridge-lg-thinq/src/platform.ts:53:7)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:4292) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
(node:4292) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

From the Homebridge GUI, it looks like the plugin correctly fetches device data but fails to interpret it dues to the device being an older version of ThinQ (even with ThinQ v1 compatibility option checked). 

I tried making a fork of the older [homebridge-wideq](https://github.com/zvuc/homebridge-wideq/commit/171b006d1a29ca0424c124834042b7caf1057526) to successfully detect the device to at least be able to send ON/OFF control signals, but due to lack of my knowledge I couldn't really make any further progress. It would be really nice if this plugin could also add support for the "legacy legacy" devices. 

I'd be happy to provide any further data if you need them.

**Comment by nVuln** [comment]
please turn on homebridge debug mode, then upload your debug logging, I need your device data wrote in debug log

**Comment by zvuc** [comment]
I actually did, but I didn't get any meaningful messages other than above referring to the Air Purifier device. (I have some other devices linked to the ThinQ app - washer and dryer, both of which are working fine & are loading from cached configs on reboot)

<img width="799" alt="image" src="https://user-images.githubusercontent.com/2870726/133931676-991bd33f-d10a-4d6b-9fb5-8384eb78be23.png">

Perhaps the Air purifier device is not showing up on log because it's somehow already registered? (But obviously not being detected properly as a compatible Air Purifier device...)

**Comment by nVuln** [comment]
you may found something like this `Found device:  {"deviceId":"d27bb3f0-7149-******",`, please upload it

**Comment by zvuc** [comment]
Actually, I think I found the reason the Air Purifier wasn't registering properly. The culprit was an older LG Air Conditioner that I had registered to ThinQ app that's not really a "Smart" device. (The device registers in the ThinQ app device list, but the only 'smart' function it has is have the AC play a fast sequence of certain beeps with different pitches while your phone hears through the microphone then match for any error codes to figure out what's wrong with the device...) So I guess that AC was blocking further device detection because it's obviously not the type of device the plugin was expecting. If you're interested in the AC device code though, here it is:

> {
    "deviceId": "EC06EEFA-1979-486D-9BBE-26F3FF1BAC58",
    "deviceType": 401,
    "modelName": "Normal-401",
    "subModelNm": null,
    "sensorType": null,
    "alias": "거실 에어컨",
    "deviceCode": "AI01",
    "networkType": "04",
    "tftYn": "N",
    "guideTypeYn": "Y",
    "guideType": "",
    "pccModelYn": "N",
    "autoOrderYn": "N",
    "drServiceYn": "N",
    "ssid": "",
    "timezoneCode": "Asia/Seoul",
    "timezoneCodeAlias": "Korea/Seoul",
    "sdsGuide": "{\"deviceCode\":\"AI01\"}",
    "newRegYn": "N",
    "remoteControlType": null,
    "fareTarget": null,
    "area": "3430012",
    "sleep": null,
    "deviceState": "D",
    "rmsClientId": null,
    "regDtUtc": "20210824003255000",
    "regIndex": 0,
    "blackboxYn": null,
    "groupableYn": "N",
    "controllableYn": "N",
    "combinedProductYn": "N",
    "masterYn": "Y",
    "manufacture": null,
    "online": false,
    "platformType": "thinq1",
    "homeDeviceOrder": -1,
    "roomDeviceOrder": 2,
    "ownershipYn": "Y",
    "modelJsonVer": "0.2",
    "modelJsonUri": "https://kic.lgthinq.com:46030/api/webContents/modelJSON?modelName=Normal-401&countryCode=WW&contentsId=JS201608230700001792&authKey=thinq",
    "appModuleVer": "17.23",
    "appModuleUri": "https://objectcontent.lgthinq.com/8cd1080c-3c86-45c5-ba15-e3e53c0f6b0d?hdnts=exp=1690436193~hmac=87f4cba3139bdbf4d52f5ed4fc15384043da8c87c0d471402e1c3eb6afb2acd1",
    "appRestartYn": "Y",
    "appModuleSize": "13511210",
    "langPackProductTypeVer": "54.1",
    "langPackProductTypeUri": "https://objectcontent.lgthinq.com/14ba044c-aa6f-4667-bddf-4737cd493bd9?hdnts=exp=1694826084~hmac=5f5a67eb9a743d06b1ed8c80cfce067c41e64ebf7fc74079d429559807e995df",
    "langPackModelVer": "",
    "langPackModelUri": "",
    "roomId": "162790180027834199",
    "fwInfoList": null,
    "modemInfo": null,
    "existsEntryPopup": "N",
    "fwVer": "",
    "modemVer": null,
    "subDeviceCount": 0,
    "firebaseLogKey": null,
    "cardType": "Small",
    "cardControl": "Base"
}


So after removing this device from the ThinQ app, removing it from the Homebridge plugin device list and rebooting, I successfully got the air purifier detected and loaded:

> [9/19/2021, 11:57:10 PM] [LGThinQ] Found device:  {"deviceId":"d27bdb00-7149-11d3-80b0-044eaf14ef16","deviceType":402,"modelName":"AIR_910604_WW","subModelNm":null,"sensorType":null,"alias":"Air Purifier","deviceCode":"LI02","networkType":"02","tftYn":"N","guideTypeYn":"Y","guideType":"AIR_TYPE1","pccModelYn":"Y","autoOrderYn":"N","drServiceYn":"N","ssid":"OikosN","timezoneCode":"Asia/Seoul","timezoneCodeAlias":"Korea/Seoul","sdsGuide":"{\"deviceCode\":\"LI02\"}","newRegYn":"N","remoteControlType":"","fareTarget":null,"area":"3430012","sleep":null,"deviceState":"E","rmsClientId":null,"regDtUtc":"20210502151111000","regIndex":0,"blackboxYn":null,"groupableYn":"N","controllableYn":"N","combinedProductYn":"N","masterYn":"Y","manufacture":null,"online":false,"platformType":"thinq1","homeDeviceOrder":1,"roomDeviceOrder":1,"ownershipYn":"Y","modelJsonVer":"3.8","modelJsonUri":"https://kic.lgthinq.com:46030/api/webContents/modelJSON?modelName=modelJSON_401&countryCode=KR&contentsId=d27bdb00-7149-11d3-80b0-044eaf14ef16&authKey=thinq","appModuleVer":"6.26","appModuleUri":"https://objectcontent.lgthinq.com/64e1799f-ebc6-40d5-be3f-a5da236eb094?hdnts=exp=1692841091~hmac=58da68817a5596b875cea58c23c970e30dae4b006a4b3d142bdfb2f1c7854743","appRestartYn":"Y","appModuleSize":"16031930","langPackProductTypeVer":"28.29","langPackProductTypeUri":"https://objectcontent.lgthinq.com/6d0beaec-f1d4-4dee-8a74-7a73b9000167?hdnts=exp=1694595693~hmac=354a114a20305cdbd087ab876f08b01a17f9cb9c61b262471d66ac028d128a26","langPackModelVer":"","langPackModelUri":"","roomId":"162687589199942252","fwInfoList":null,"modemInfo":null,"existsEntryPopup":"N","fwVer":"2.6.2_RTOS_3K","modemVer":null,"subDeviceCount":0,"firebaseLogKey":"P:MP","cardType":"Small","cardControl":"Base"}
[9/19/2021, 11:57:10 PM] [LGThinQ] [d27bdb00-7149-11d3-80b0-044eaf14ef16] Device model cache missed.
[9/19/2021, 11:57:10 PM] [LGThinQ] Adding new accessory: Air Purifier - d27bdb00-7149-11d3-80b0-044eaf14ef16

The power on/off switches seem to work, but the light on/off seemingly not (I don't think my device has support for it because I don't have hardware buttons on the device nor similar option inside the ThinQ app)

But the fan control also don't work. I am kind of guessing that the fan speed translation is incorrect, seeing that the fan speed is at 4% when it's at maximum level?

<img width="1088" alt="image" src="https://user-images.githubusercontent.com/2870726/133932649-58b6de78-12a6-48f6-a6cf-a9c0f2e49626.png">

**Comment by nVuln** [comment]
@zvuc thank you, I'll handle that "stupid" AC 😵‍💫
for fan speed, currently it have 4 level of speed, level 1 -> 4, mapped to low, medium, high and extra, and it's scale to 0-100% in homekit
look like homebridge don't scale down it even I registered RotationSpeed maxValue = 4

for light control, I'll make a config option to enable/disable it

**Comment by zvuc** [comment]
@nVuln I honestly don't think you have to go through the hassle of handling that AC... it's pretty much useless in the ThinQ app anyways and I should have removed it long ago anyways. 😂

For the fan speed btw, my Air Purifier has an "Auto" option too, which is where I usually put my device at daytime — does the current implementation support that value? If not, do you think we can maybe map it to 0% fan speed? I'm not even sure if other newer LG air purifiers with finer fan speed controls allow completely turning off fan speed without turning off the entire device... 🤔

**Comment by nVuln** [comment]
@zvuc Air purifier have Auto/Manual mode, please try turn on Auto and see if it's worked for you, I'm not sure about this because they have multiple OpMode value and it's different on each device model 😵‍💫

**Comment by zvuc** [comment]
@nVuln where/how do I turn on Auto? I don't see any relevant options or controls on the homebridge side. 🤔

**Comment by zvuc** [comment]
BTW: I don't think it's relevant to the Fan speed issue but I get an error message regarding Air quality meters:
> [9/20/2021, 12:53:11 AM] [homebridge-lg-thinq] This plugin generated a warning from the characteristic 'Air Quality': characteristic value expected valid finite number and received "NaN" (number). See https://git.io/JtMGR for more info.
[9/20/2021, 12:53:11 AM] [homebridge-lg-thinq] Error: 
    at AirQuality.Characteristic.characteristicWarning (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Characteristic.ts:2038:105)
    at AirQuality.Characteristic.validateUserInput (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Characteristic.ts:1931:16)
    at AirQuality.Characteristic.updateValue (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Characteristic.ts:1379:20)
    at AirQualitySensor.Service.updateCharacteristic (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Service.ts:465:35)
    at AirPurifier.updateAccessoryCharacteristic (/homebridge/node_modules/homebridge-lg-thinq/src/devices/AirPurifier.ts:182:28)
    at AirPurifier.update (/homebridge/node_modules/homebridge-lg-thinq/src/baseDevice.ts:32:10)
    at EventEmitter.emit (events.js:400:28)
    at Timeout._onTimeout (/homebridge/node_modules/homebridge-lg-thinq/src/platform.ts:192:25)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)

**Comment by nVuln** [comment]
> @nVuln where/how do I turn on Auto? I don't see any relevant options or controls on the homebridge side. 🤔

In Home app

![image](https://user-images.githubusercontent.com/54855446/133934282-410d07b4-f416-48e7-b7a1-1a79b907d27f.jpeg)

**Comment by zvuc** [comment]
Yeah I don't think it's showing up on mine. 😞

<img width="522" alt="image" src="https://user-images.githubusercontent.com/2870726/133934611-3cf52ff2-882e-4ae1-bb5b-4f4d5e60cb2f.png">

Observing the json file from the link printed in the device found log, ( https://kic.lgthinq.com:46030/api/webContents/modelJSON?modelName=modelJSON_401&countryCode=KR&contentsId=d27bdb00-7149-11d3-80b0-044eaf14ef16&authKey=thinq ) I think my device actually has "Auto" mode at value `9` 🤔 

> "SupportWindStrength": {
      "_comment": "FuncSync_�ٶ�����",
      "type": "Enum",
      "option": {
        "0": "@NON",
        "3": "@AP_MAIN_MID_WINDSTRENGTH_LOW_W",
        "5": "@AP_MAIN_MID_WINDSTRENGTH_MID_W",
        "7": "@AP_MAIN_MID_WINDSTRENGTH_HIGH_W",
        "9": "@AP_MAIN_MID_WINDSTRENGTH_AUTO_W"
      }
    },


Also btw the fan control just plain doesnt work at the moment regardless of auto/manual switch. 😂

**Comment by nVuln** [comment]
> BTW: I don't think it's relevant to the Fan speed issue but I get an error message regarding Air quality meters:

I don't have Air purifier thinq1 device, so I cannot debug it.
can you put some debug code around here and see what happen: [src/v1/transforms/AirState.ts](https://github.com/nVuln/homebridge-lg-thinq/blob/master/src/v1/transforms/AirState.ts) , the concept of this code is transform from thinq1 monitor device to thinq2 device data, and process it and single place

I saw your device have two variable `TotalAirPolution` and `AirPolution`, don't know which one is correct

> Yeah I don't think it's showing up on mine. 😞

select Accessories tab, and select Air Purififer accessory, you can see manual/auto mode there

> Also btw the fan control just plain doesnt work at the moment regardless of auto/manual switch. 😂
> Nothing changes on the actual device regardless of where i put the fan slider from HomeKit app.

I don't know if you can switch WindStrength to `9`, it's based on `OpMode` value
example in my model, if `OpMode` = 16 mean @AP_MAIN_MID_OPMODE_AUTO_W, I can't control windstrength

**Comment by zvuc** [comment]
Okay just realized the difference in the screenshot was caused by the accessories (Air Purifier, "Lights", Sensor) being merged into one.

But even after splitting them into individual devices, the fan controls don't seem to actually change anything on the device.

<img width="539" alt="image" src="https://user-images.githubusercontent.com/2870726/133935392-2587d022-913c-49b5-9f9a-f678c994674d.png">

Also something strange: turning off the device does seem to send an OFF signal, but in the homebridge and Apple Home app, the accessory button turns itself on again (not the actual device). Meanwhile the "Lights" accessory stays off.

My educated guess: my device has an 'Sensor always On' option you can set in the ThinQ app where if it's on, the device actually stays on to keep the sensors running (The LEDs, displays go off, fans spin down, but it's somehow still on). When that option is toggled off, turning off the device powers down HARD meaning when you turn it back on again, the sensors have to re-calibrate before displaying numbers (it shows a 'spinning' animation.)

FYI: how the "display" on my device looks
<img width="542" alt="image" src="https://user-images.githubusercontent.com/2870726/133935566-b9f8434b-072d-4e21-86d3-98d5f058e773.png">

**Comment by nVuln** [comment]
Please make some debug in your side and see OpMode value in auto mode and other mode

**Comment by nVuln** [comment]
are you sure your device don't have light control feature ? even in LG thinQ native app ?
my device don't have physical light button, but I still control it via LG app

**Comment by zvuc** [comment]
yeah, mine really doesn't have that option. 
![image](https://user-images.githubusercontent.com/2870726/134015009-06988ee0-2b85-42a0-a8d5-6b1c558871ae.png)

Although, the device does have a "sleep timer" mode: which turns off all LEDs and display except the "Reservation" LED, turns down fan speed to the lowest level, and turns off after 7 hours. Not sure if that's translated to the 'SignalLighting' in the internal status...

**Comment by zvuc** [comment]
Hey, after updating to the latest version (1.2.6), I get this message in console (I think it's new?)

> [9/21/2021, 4:38:14 PM] [LGThinQ] [request] post https://kic.lgthinq.com:46030/api/rti/rtiResult
[9/21/2021, 4:38:14 PM] [LGThinQ] [Air Purifier] Received snapshot:  {"airState.opMode":9,"airState.operation":true,"airState.tempState.current":40,"airState.tempState.target":0,"airState.windStrength":8,"airState.wDir.vStep":0,"airState.wDir.hStep":0,"airState.circulate.rotate":0,"airState.lightingState.signal":1,"airState.quality.overall":1,"airState.quality.sensorMon":0,"airState.quality.PM1":8,"airState.quality.PM2":8,"airState.quality.PM10":8,"online":true,"raw":{"Operation":"1","OpMode":"9","WindStrength":"8","TempUnit":"NS","TempCur":"40","TempCfg":"0","GroupType":"12","SleepTime":"0","OnTime":"0","OffTime":"0","RacAddFunc":"NS","ExtraOp":"0","DiagCode":"00","TimeBsOn":"0","TimeBsOff":"0","AirClean":"0","AutoDry":"0","PowerSave":"0","WDirVStep":"0","WDirHStep":"0","TempLimitMax":"0","TempLimitMin":"0","DuctZoneType":"0","ZoneControl":"0","DRED":"0","SensorPM1":"8","SensorPM2":"8","SensorPM10":"8","AirPolution":"1","HumidityCfg":"0","WaterTempCoolMin":"NS","WaterTempCoolMax":"NS","WaterTempHeatMin":"NS","WaterTempHeatMax":"NS","HotWaterTempMin":"NS","HotWaterTempMax":"NS","SensorHumidity":"0","TotalAirPolution":"1","SensorMon":"0","CleanDry":"0","ProductStatus":"0","AirMonitoring":"0","Humidification":"0","AirFast":"0","AirRemoval":"0","AirUVDisinfection":"0","WatertankLight":"0","SignalLighting":"1","WDirUpDown":"0","WDirLeftRight":"0","WSwirl":"0","Jet":"0","LowHeating":"0","CirculateStrength":"0","CirculateDir":"0","AntiBugs":"0","IceValley":"0","Humsave":"0","WaterTempCur":"0","HotWaterTempCur":"0","HotWaterTempCfg":"0","HotWaterMode":"0","HotWater":"0","AWHPTempCfgSwitch":"0","AirTempCoolMin":"NS","AirTempCoolMax":"NS","AirTempHeatMin":"NS","AirTempHeatMax":"NS","WaterInTempCur":"0","AWHPWATempControlSta":"0","DisplayControl":"0","SmartCare":"0","TwoSetCoolTemp":"0","TwoSetHeatTemp":"0","TwoSetCoolUSL":"0","TwoSetCoolLSL":"0","TwoSetHeatUSL":"0","TwoSetHeatLSL":"0","TwoSetACOState":"0","TwoSetModeDeadband":"0","TwoSetState":"0"}}

**Comment by nVuln** [comment]
yep, look like your device do not support light control via API, but light status still showing correct
can you check `OpMode` value when Auto mode is on ?

**Comment by zvuc** [comment]
I tried that earlier too, but seemed like `opMode` doesnt change — it stays at `9`.

The `airState.windStrength` does change though, Auto was `8`, Low (manual) was `2`, Mid is `4`, High is `6`.

There is also a "Turbo" mode though, which is a temporary MAX fan speed mode that returns to AUTO mode after a set timer. It sets `windStrength` to a whopping `13`. (Seems like it's not out of 10!)

This is the log message when I turn on Turbo: 
> [Air Purifier] Received snapshot:  {"airState.opMode":9,"airState.operation":true,"airState.tempState.current":40,"airState.tempState.target":0,"airState.windStrength":13,"airState.wDir.vStep":0,"airState.wDir.hStep":0,"airState.circulate.rotate":0,"airState.lightingState.signal":1,"airState.quality.overall":1,"airState.quality.sensorMon":0,"airState.quality.PM1":8,"airState.quality.PM2":8,"airState.quality.PM10":9,"online":true,"raw":{"Operation":"1","OpMode":"9","WindStrength":"13","TempUnit":"NS","TempCur":"40","TempCfg":"0","GroupType":"12","SleepTime":"0","OnTime":"0","OffTime":"0","RacAddFunc":"NS","ExtraOp":"0","DiagCode":"00","TimeBsOn":"0","TimeBsOff":"0","AirClean":"0","AutoDry":"0","PowerSave":"0","WDirVStep":"0","WDirHStep":"0","TempLimitMax":"0","TempLimitMin":"0","DuctZoneType":"0","ZoneControl":"0","DRED":"0","SensorPM1":"8","SensorPM2":"8","SensorPM10":"9","AirPolution":"1","HumidityCfg":"0","WaterTempCoolMin":"NS","WaterTempCoolMax":"NS","WaterTempHeatMin":"NS","WaterTempHeatMax":"NS","HotWaterTempMin":"NS","HotWaterTempMax":"NS","SensorHumidity":"0","TotalAirPolution":"1","SensorMon":"0","CleanDry":"0","ProductStatus":"0","AirMonitoring":"0","Humidification":"0","AirFast":"1","AirRemoval":"0","AirUVDisinfection":"0","WatertankLight":"0","SignalLighting":"1","WDirUpDown":"0","WDirLeftRight":"0","WSwirl":"0","Jet":"0","LowHeating":"0","CirculateStrength":"0","CirculateDir":"0","AntiBugs":"0","IceValley":"0","Humsave":"0","WaterTempCur":"0","HotWaterTempCur":"0","HotWaterTempCfg":"0","HotWaterMode":"0","HotWater":"0","AWHPTempCfgSwitch":"0","AirTempCoolMin":"NS","AirTempCoolMax":"NS","AirTempHeatMin":"NS","AirTempHeatMax":"NS","WaterInTempCur":"0","AWHPWATempControlSta":"0","DisplayControl":"0","SmartCare":"0","TwoSetCoolTemp":"0","TwoSetHeatTemp":"0","TwoSetCoolUSL":"0","TwoSetCoolLSL":"0","TwoSetHeatUSL":"0","TwoSetHeatLSL":"0","TwoSetACOState":"0","TwoSetModeDeadband":"0","TwoSetState":"0"}}

**Comment by zvuc** [comment]
I think I maybe found the culprit to the Purifier device button in HomeKit turning itself on even when the actual device is off: 

The `airState.operation` stays as `true` even when I turn off the device. The `"Operation"` in the raw list does change from 1 to 0. Maybe this is the value we should reference to when determining device power state?

**Comment by nVuln** [comment]
I already published test version `v1.2.7-test.2`, fixed operation status, auto mode based on windstrength, when turn on auto, plugin will send command windstrength = `8`
please try it

**Comment by zvuc** [comment]
Thanks for the work!
I've tested the new test version right away. Sharing what I've observed so far:

## 1
I occasionally get this error in homebridge log when I try to send remote control signals through HomeKit:

> [9/23/2021, 1:52:09 PM] [LGThinQ] Unknown Error:  TokenError: 0011 - 등록되지 않은 모델입니다.
    at client.interceptors.response.use.err.response._a (/homebridge/node_modules/homebridge-lg-thinq/src/lib/request.ts:31:13)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at API.request (/homebridge/node_modules/homebridge-lg-thinq/src/lib/API.ts:53:12)
    at API.postRequest (/homebridge/node_modules/homebridge-lg-thinq/src/lib/API.ts:42:12)
    at API.thinq1PostRequest (/homebridge/node_modules/homebridge-lg-thinq/src/lib/API.ts:245:12)
    at AIR_910604_WW.setActive (/homebridge/node_modules/homebridge-lg-thinq/src/v1/devices/AirPurifier.ts:8:5)

(fyi 등록되지 않은 모델입니다. is "Not a registered model" in korean)
I think this error pops when when I have the ThinQ app running on my phone. Similarly, I can't access the Air Purifier item from my ThinQ app saying it's "In Use" when I meddle with the device in HomeKit.

## 2
<img width="517" alt="image" src="https://user-images.githubusercontent.com/2870726/134456658-b6038dcb-f32c-42c3-960c-6f51267432bf.png">
Setting to "Manual" mode seems to set `airState.windStrength` to `7`, which is not a valid value for my device. (2,4,6,8 being valid values) so the device shows no LED lit up for wind speed (Auto, Low, Mid, High)

## 3
I've pressed each buttons for wind speed + turbo on the device to see if correctly reflect state inside the HomeKit app.
![purifier](https://user-images.githubusercontent.com/2870726/134457662-48c78273-0b05-43b5-a76b-eb79dd1c905c.png)
It does seem the device state now gets updated when fan speed changes, but I have some suggestions:
- Slider probably should not show in AUTO mode. There is no way to tell difference between Turbo and Auto.
- LOW should not be showing at 0% level. HIGH should probably be showing at 100% level. LOW being displayed as 0% fan speed causes a problem where if I turn down the fan speed in HomeKit app, it just sends an OFF signal instead.
- Turbo should probably be another optional switch down there because it acts as a special function with auto off timer.
- There are two sliders shown for the fan speed, one at the top (vertical) and one at the bottom (horizontal). If the first one does work correctly, I don't think we need to show the second one.

## 4
![Artboard](https://user-images.githubusercontent.com/2870726/134457976-a7f0721d-4d42-4028-b6f3-621ea09a44df.png)
"Oscillate" seems to be an option for some air purifier models that support "swiveling", but mine doesn't support that.
Also, there is an additional "Mode" switch that seems to just turn off the device when I switch it off.

**Comment by nVuln** [comment]
`v1.2.7-test.3` is published now

> (fyi 등록되지 않은 모델입니다. is "Not a registered model" in korean)

handled it

> Setting to "Manual" mode seems to set airState.windStrength to 7, which is not a valid value for my device. (2,4,6,8 being valid values) so the device shows no LED lit up for wind speed (Auto, Low, Mid, High)

removed windStrength `7` also

> Slider probably should not show in AUTO mode. There is no way to tell difference between Turbo and Auto.

I can't hide it, it's Home app

> LOW should not be showing at 0% level. HIGH should probably be showing at 100% level. LOW being displayed as 0% fan speed causes a problem where if I turn down the fan speed in HomeKit app, it just sends an OFF signal instead.

fixed it, I changed minStep from `1` to `0.1`, make sure you can't scroll down to 0% and trigger OFF signal

> There are two sliders shown for the fan speed, one at the top (vertical) and one at the bottom (horizontal). If the first one does work correctly, I don't think we need to show the second one.

I can't hide it

> "Oscillate" seems to be an option for some air purifier models that support "swiveling", but mine doesn't support that. It should be removed for models that don't support it.

removed swingMode on your model also

> There is an additional "Mode" switch that seems to just turn off the device when I switch it off. It probably should be removed too.

I'm updating to iPadOS 15 too, let see what button is this, I don't see it on ipados 14

**Comment by zvuc** [comment]
Just updated to test.3 version.

Getting some new errors now:

> [9/23/2021, 5:28:29 PM] [homebridge-lg-thinq] This plugin generated a warning from the characteristic 'Air Quality': characteristic value expected valid finite number and received "NaN" (number). See https://git.io/JtMGR for more info.
[9/23/2021, 5:28:29 PM] [homebridge-lg-thinq] Error: 
    at AirQuality.Characteristic.characteristicWarning (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Characteristic.ts:2038:105)
    at AirQuality.Characteristic.validateUserInput (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Characteristic.ts:1931:16)
    at AirQuality.Characteristic.updateValue (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Characteristic.ts:1379:20)
    at AirQualitySensor.Service.updateCharacteristic (/usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/src/lib/Service.ts:465:35)
    at AIR_910604_WW.updateAccessoryCharacteristic (/homebridge/node_modules/homebridge-lg-thinq/src/models/AIR_910604_WW.ts:76:28)
    at new AirPurifier (/homebridge/node_modules/homebridge-lg-thinq/src/devices/AirPurifier.ts:72:10)
    at new AirPurifier (/homebridge/node_modules/homebridge-lg-thinq/src/v1/devices/AirPurifier.ts:5:1)
    at new AIR_910604_WW (/homebridge/node_modules/homebridge-lg-thinq/src/models/AIR_910604_WW.ts:24:5)
    at LGThinQHomebridgePlatform.discoverDevices (/homebridge/node_modules/homebridge-lg-thinq/src/platform.ts:144:25)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
[9/23/2021, 5:28:29 PM] [LGThinQ] Cannot read property 'updateCharacteristic' of undefined
[9/23/2021, 5:28:29 PM] [LGThinQ] TypeError: Cannot read property 'updateCharacteristic' of undefined
    at AIR_910604_WW.updateAccessoryCharacteristic (/homebridge/node_modules/homebridge-lg-thinq/src/models/AIR_910604_WW.ts:78:27)
    at new AirPurifier (/homebridge/node_modules/homebridge-lg-thinq/src/devices/AirPurifier.ts:72:10)
    at new AirPurifier (/homebridge/node_modules/homebridge-lg-thinq/src/v1/devices/AirPurifier.ts:5:1)
    at new AIR_910604_WW (/homebridge/node_modules/homebridge-lg-thinq/src/models/AIR_910604_WW.ts:24:5)
    at LGThinQHomebridgePlatform.discoverDevices (/homebridge/node_modules/homebridge-lg-thinq/src/platform.ts:144:25)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)

**Comment by nVuln** [comment]
please try `v1.2.7-test.4`

**Comment by zvuc** [comment]
I installed the `1.2.7-test.4` version.
- Turbo seems to be working correctly. (It shows as a separate smart 'button' device but I can group them with the main device alright)
- 'Low' fan speed still seems to be going down to 0 visually, although now it doesn't trigger off.
- When in Auto mode, adjusting fan speed through large slider (the blue vertical one) sends a signal (I hear the beep sound from my device) but doesn't seem to change anything.
- Generally, dragging the large blue Fan speed slider seems to be buggy -- I hear multiple beep sounds and see multiple POST messages on log, but many of the times the slider returns to prior state when fetch occurs (I have set it to 5 seconds now for testing). I'm guessing that the slider can send inbetween values when I'm touch holding the slider and that causes some confusion. I'm not entirely sure but is there an 'incremental' variant of the slider so that they can be only set to preset values? (LOW:2 = 25%, MID:4 = 50%, HIGH: 75%, AUTO: 100%)

I'm attaching a screencast I took where I try to manipulate a few controls, for reference.

https://user-images.githubusercontent.com/2870726/134768840-ebf9e612-f5e0-41d4-8286-6e19f277c000.MOV
